Refactor GlusterFS PV spec.

humblec opened the pull request:
Refactor GlusterFS PV spec.
    
    This patch introduces glusterfsPersistentVolumeSource addition
    to glusterfsVolumeSource. All fields remains same as glusterfsVolumeSource
    with an addition of a new field
    called `EndpointsNamespace` to define namespace of endpoint in the
    spec.
    


Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->


**Special notes for your reviewer**:




**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
GlusterFS PersistentVolumes sources can now reference endpoints in any namespace using the `spec.glusterfs.endpointsNamespace` field. Ensure all kubelets are upgraded to 1.13+ before using this capability.
```

humblec commented:
/test pull-kubernetes-bazel-test

humblec commented:
/test pull-kubernetes-verify

humblec commented:
@rootfs

humblec commented:
/retest

rootfs commented:
@humblec can you rebase?

rootfs commented:
cc @liggitt

liggitt commented on pkg/api/persistentvolume/util.go:
```diff
@@ -127,6 +127,18 @@ func VisitPVSecretNames(pv *api.PersistentVolume, visitor Visitor) bool {
+				// use the endpoint namespace if namespace is set
+				ns = *source.Glusterfs.EndpointsNamespace
+			}
+			if !visitor(ns, *source.Glusterfs.EndpointsNamespace, true /* kubeletVisible */) {
```
why are we passing an endpoints namespace twice to a function designed to get the named of referenced secrets?

humblec replied:
iiuc, this func ( `visitpvsecret`) entry is not required at all. I have commented out this part in the latest patch set. May be I am missing something here, please let me know if thats the case.

liggitt replied:
correct, this change should be dropped... the field you are adding is not a reference to a secret, so there should be no changes here

liggitt commented on pkg/apis/core/types.go:
```diff
@@ -914,6 +914,28 @@ type GlusterfsVolumeSource struct {
+	ReadOnly bool
+
+	// EndpointsNameSpace is the endpoint namespace that contains Glusterfs endpoint.
+	// More info: https://releases.k8s.io/HEAD/examples/volumes/glusterfs/README.md#create-a-pod
```
document behavior if empty (defaults to the same namespace as the bound PVC, right?)

humblec replied:
Done.

liggitt commented on pkg/printers/internalversion/describe.go:
```diff
@@ -893,6 +893,15 @@ func printGlusterfsVolumeSource(glusterfs *api.GlusterfsVolumeSource, w PrefixWr
+		"    EndpointsNamespace:\t%v\n"+
+		"    Path:\t%v\n"+
+		"    ReadOnly:\t%v\n",
+		glusterfs.EndpointsName, glusterfs.EndpointsNamespace, glusterfs.Path, glusterfs.ReadOnly)
```
do you want to  print `nil` for EndpointsNamespace or show the PVC namespace?

humblec replied:
I was thinking  that, it can be nil or a namespace which is in the spec. Are you saying its going to be `nil` always ?

liggitt replied:
it's nil by default... I guess we don't have the namespace of the pvc here to display in that case

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -160,8 +158,33 @@ func (plugin *glusterfsPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volu
+		endpoints := spec.PersistentVolume.Spec.Glusterfs.EndpointsName
+		endpointsNs := defaultNamespace
+		if spec.PersistentVolume.Spec.Glusterfs.EndpointsNamespace != nil {
+			endpointsNs := spec.PersistentVolume.Spec.Glusterfs.EndpointsNamespace
```
validation must ensure if EndpointsNamespace is non-nil, it is non-empty, otherwise this will return "" as the namespace

humblec replied:
Done.

humblec commented:
@rootfs rebased, thanks!

humblec commented:
@liggitt thanks for the review comments, hopefully I have addressed all. Please take a look.

humblec commented on pkg/apis/core/validation/validation.go:
```diff
@@ -928,6 +928,16 @@ func validateGlusterfsVolumeSource(glusterfs *core.GlusterfsVolumeSource, fldPat
+	allErrs := field.ErrorList{}
+	if len(glusterfs.EndpointsName) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("endpoints"), ""))
+	}
```
Done.

humblec commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -621,6 +621,28 @@ type GlusterfsVolumeSource struct {
+
+	// EndpointsNameSpace is the endpoint namespace that contains Glusterfs endpoint.
+	// More info: https://releases.k8s.io/HEAD/examples/volumes/glusterfs/README.md#create-a-pod
+	EndpointsNamespace *string `json:"endpointsnamespace" protobuf:"bytes,4,opt,name=endpointsnamespace"`
```
Done..

liggitt commented on pkg/apis/core/types.go:
```diff
@@ -921,6 +921,28 @@ type GlusterfsVolumeSource struct {
+	// +optional
+	ReadOnly bool
+
+	// EndpointsNameSpace is the endpoint namespace that contains Glusterfs endpoint.
```
match case on the field name

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -902,6 +902,23 @@ func validateGlusterfsVolumeSource(glusterfs *core.GlusterfsVolumeSource, fldPat
+		if len(*endpointNs) == 0 {
+			allErrs = append(allErrs, field.Required(fldPath.Child("endpointsNamespace"), ""))
+		}
+		allErrs = append(allErrs, ValidateDNS1123Label(*endpointNs, fldPath.Child("endpointsNamespace"))...)
```
don't run the format check if the length is 0 and we already added a "required" error

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -160,8 +158,36 @@ func (plugin *glusterfsPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volu
+				return endpoints, endpointsNs, nil
+			}
+		}
+		return "", endpointsNs, nil
```
should be `return endpoints, endpointsNs, nil`, right?

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -160,8 +158,36 @@ func (plugin *glusterfsPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volu
+		spec.PersistentVolume.Spec.Glusterfs != nil {
+		endpoints := spec.PersistentVolume.Spec.Glusterfs.EndpointsName
+		endpointsNs := defaultNamespace
+		if spec.PersistentVolume.Spec.Glusterfs.EndpointsNamespace != nil {
```
simplify to
```
if overriddenNs := spec.PersistentVolume.Spec.Glusterfs.EndpointsNamespace; overriddenNs != nil && len(*overriddenNs) > 0 {
  endpointsNs = overriddenNs
}
```

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -98,15 +98,15 @@ func (plugin *glusterfsPlugin) GetPluginName() string {
 
 	return fmt.Sprintf(
 		"%v:%v",
-		volumeSource.EndpointsName,
```
what are the uniqueness requirements around what GetVolumeName returns today? why is the namespace not included already?

liggitt replied:
still outstanding

liggitt replied:
this question is still outstanding. what is GetVolumeName used for, and why does it not include the endpoint's namespace today?

humblec replied:
@liggit apologies for missing this comment.. Here is my thought,  `path` ( ie volume name) is unique for a gluster cluster. That said, only one volume exist in a gluster cluster, so its unique in nature.  But, there is a possibility exist that, second gluster cluster can have the same volume name. Endpoint basically carries gluster nodes IPs , so the uniqueness can be achieved by a combo of endpoint and path.

liggitt replied:
> Endpoint basically carries gluster nodes IPs , so the uniqueness can be achieved by a combo of endpoint and path.

two endpoints with the same name in two different namespaces could reference different IPs. 

example:
* ns1 has endpoint named "gluster" pointing to IP 1.1.1.1, and path `/path/to/volume`
* ns2 has endpoint named "gluster" pointing to IP 2.2.2.2, and path `/path/to/volume`

this method would compute a volume name of "gluster:/path/to/volume" for both, right?

like I said, this isn't new to this PR, but it needs further explanation as to why this is ok. is the namespace information included somehow in the computed path? if so, which namespace is considered, the namespace of the consuming pod, or the namespace of the gluster endpoint?

humblec replied:
> ns1 has endpoint named "gluster" pointing to IP 1.1.1.1, and path /path/to/volume
    ns2 has endpoint named "gluster" pointing to IP 2.2.2.2, and path /path/to/volume>

Sure.

>this method would compute a volume name of gluster:/path/to/volume for both, right?>

@liggitt, The compute mount will be `1.1.1.1:/path/to/volume` for ns1 and `2.2.2.2:/path/to/volume` for ns2, as we use IPs in endpoint for mount.  I am not 100% sure that,  compute mount path is the only scenario we have to think about though, but survived till now :).

liggitt replied:
> The compute mount will be `1.1.1.1:/path/to/volume` for ns1 and `2.2.2.2:/path/to/volume` for ns2, as we use IPs in endpoint for mount. I am not 100% sure that, compute mount path is the only scenario we have to think about though, but survived till now :).

That's not what this code does. If the endpoint IP gets resolved elsewhere, that's fine, but I'm concerned that GetVolumeName is relying on some other code that is implicitly making use of the pod's namespace to disambiguate the mounts. With the change this PR is making, the pod's namespace is no longer the correct disambiguator in all cases.

humblec replied:
Sure.  @liggitt  pushing `endpointnamespace` with `endpoint and path` if the persistentvolume spec is non nil in return will do ?

liggitt replied:
`endpointsName:endpointsNamespace:volumePath` (only when pv endpoints namespace is non-nil for backcompat) would be unique and would resolve my concerns about this method. would need an ack from @kubernetes/sig-storage-pr-reviews that that doesn't break assumptions elsewhere (does anything try to parse out endpointsName and path from this volume name elsewhere?)

humblec replied:
> endpointsName:endpointsNamespace:volumePath (only when pv endpoints namespace is non-nil for backcompat) would be unique and would resolve my concerns about this method>

Sure, will address it in next patch set.

>would need an ack from @kubernetes/sig-storage-pr-reviews that that doesn't break assumptions elsewhere (does anything try to parse out endpointsName and path from this volume name elsewhere?)>

afaik, its not parsed anywhere. the closest code is  'reconstructvolume' , but reconstuction is not supported (https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/glusterfs/glusterfs.go#L199) for glusterfs. `GetVolumeName()` is mainly used for `attach/detach` volumes and glusterfs is not into this category.  @jsafrane @rootfs can you confirm my thoughts  or share details?

humblec replied:
@liggitt added endpointsnamespace as mentioned. PTAL :) .. Thanks !!!

liggitt replied:
cc @msau42 on modification to GetVolumeName()

msau42 replied:
GetVolumeName() is used in a few places, all of which are not applicable to gluster:
* AD controller
* kubelet volume manager actual/desired sw for attachable volumes
* volume reconstruction

Can we just return error from GetVolumeName() since it shouldn't even be called?

msau42 replied:
Since this function should not come into play for gluster, backwards compatibility should not be a concern.  I am fine with cleaning this up in a followup.

humblec commented:
@liggitt back from bit long medical leaves :(, I will go through the review comments are address them soon. Apologies for the delay.

humblec commented:
Comments are addressed, however `make update` fails with below error message. 
```
    ./vendor/k8s.io/kube-openapi/cmd/openapi-gen
2018/07/19 23:35:42 Code for OpenAPI definitions generated
7d6
< API rule violation: names_match,k8s.io/api/core/v1,GlusterfsPersistentVolumeSource,EndpointsName
8a8
> API rule violation: names_match,k8s.io/api/core/v1,GlusterfsPersistentVolumeSource,EndpointsName
Error: API rules check failed. Reported violations "_output/violations.report" differ from known violations "hack/testdata/api-rules/violation_exceptions.list". Please fix API source file if new violation is detected, or update known violations "hack/testdata/api-rules/violation_exceptions.list" if existing violation is being fixed. Please refer to hack/testdata/api-rules/README.md and https://github.com/kubernetes/kube-openapi/tree/master/pkg/generators/rules for more information about the API rules being enforced.
Makefile.generated_files:445: recipe for target 'pkg/generated/openapi/zz_generated.openapi.go' failed
make[1]: *** [pkg/generated/openapi/zz_generated.openapi.go] Error 1
make[1]: Leaving directory '/go/src/k8s.io/kubernetes'
Makefile:483: recipe for target 'generated_files' failed
make: *** [generated_files] Error 2
make: Leaving directory '/go/src/k8s.io/kubernetes'
!!! [0719 23:35:42] Call tree:
!!! [0719 23:35:42]  1: /home/hchiramm/gopath/src/k8s.io/kubernetes/hack/../build/../build/common.sh:553 kube::build::run_build_command_ex(...)
!!! [0719 23:35:42]  2: /home/hchiramm/gopath/src/k8s.io/kubernetes/hack/../build/run.sh:39 kube::build::run_build_command(...)
Running update-generated-protobuf FAILED
make: *** [Makefile:158: update] Error 1
```

Waiting for @liggitt's ( expert's :) ) opinion on above tool check failure, once this is sorted, I will generate the spec files.

liggitt commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -624,6 +624,29 @@ type GlusterfsVolumeSource struct {
+type GlusterfsPersistentVolumeSource struct {
+	// EndpointsName is the endpoint name that details Glusterfs topology.
+	// More info: https://releases.k8s.io/HEAD/examples/volumes/glusterfs/README.md#create-a-pod
+	EndpointsName string `json:"EndpointsName" protobuf:"bytes,1,opt,name=EndpointsName"`
```
this must be a copy/paste of the field from GlusterfsVolumeSource

liggitt replied:
```
EndpointsName string `json:"endpoints" protobuf:"bytes,1,opt,name=endpoints"`
```

humblec replied:
Sure, it was experimented in latest patch set to get rid of violation , now reverted the change back to reflect above.

liggitt commented on hack/testdata/api-rules/violation_exceptions.list:
```diff
@@ -5,6 +5,7 @@ API rule violation: names_match,k8s.io/api/core/v1,DaemonEndpoint,Port
 API rule violation: names_match,k8s.io/api/core/v1,Event,ReportingController
 API rule violation: names_match,k8s.io/api/core/v1,FCVolumeSource,WWIDs
 API rule violation: names_match,k8s.io/api/core/v1,GlusterfsVolumeSource,EndpointsName
+API rule violation: names_match,k8s.io/api/core/v1,GlusterfsPersistentVolumeSource,EndpointsName
```
I think this file is required to be alphabetical, so move this up one line

liggitt replied:
@roycaihw fyi on usability of the generation error message ^

roycaihw replied:
Acknowledge. I'll add sorting before comparing the violations

humblec replied:
@liggitt Thanks!! yeah, that made the `make update` to work at my end!

roycaihw commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -624,6 +624,29 @@ type GlusterfsVolumeSource struct {
+type GlusterfsPersistentVolumeSource struct {
+	// EndpointsName is the endpoint name that details Glusterfs topology.
+	// More info: https://releases.k8s.io/HEAD/examples/volumes/glusterfs/README.md#create-a-pod
+	EndpointsName string `json:"endpoints" protobuf:"bytes,1,opt,name=endpoints"`
```
IIUC the fields name are kept the same as `GlusterfsVolumeSource` for some sort of backwards compatibility and symmetry. But it would be backwards incompatible if someone expects `PersistentVolumeSource.Glusterfs` to be `GlusterfsVolumeSource` (which turns out to be `GlusterfsPersistentVolumeSource` now). 

In that case should we fix the go name (s/EndpointsName/Endpoints) now to follow the api rule https://github.com/kubernetes/kubernetes/blob/master/vendor/k8s.io/kube-openapi/pkg/generators/rules/names_match.go#L53-L78? 

Also FYI I'm planning on fixing the existing go name violations (e.g. https://github.com/kubernetes/kubernetes/pull/66097). The fix will be breaking change, but it's mechanical change and will be caught by compiler.

liggitt replied:
> But it would be backwards incompatible if someone expects PersistentVolumeSource.Glusterfs to be GlusterfsVolumeSource (which turns out to be GlusterfsPersistentVolumeSource now).

Not via the API. The serialized JSON and protobuf output is identical, no matter what struct it comes from.

> In that case should we fix the go name (s/EndpointsName/Endpoints) now to follow the api rule https://github.com/kubernetes/kubernetes/blob/master/vendor/k8s.io/kube-openapi/pkg/generators/rules/names_match.go#L53-L78?

No, we must maintain wire compatibility with all the fields of the old struct.

> Also FYI I'm planning on fixing the existing go name violations (e.g. #66097)

That is ok as long as the serialized wire formats do not change in any way.

roycaihw replied:
Yes the change is backwards compatible via the API. I'm thinking if someone builds a go client that manipulates the field `Glusterfs` as a `GlusterfsVolumeSource` type, e.g.
```
    pvs := PersistentVolumeSource{
        Glusterfs: &GlusterfsVolumeSource{}
    }
```
then it will be a breaking change when the user updates `k8s.io/api/core/v1` package, because `Glusterfs` should be of type `GlusterfsPersistentVolumeSource` now. Thus I'm thinking we could also fix the go field name `EndpointsName` to be `Endpoints` as part of the breaking change. Sorry I guess my previous comment was misleading, I agree we shouldn't change the json field name (the serialized wire formats), but we could change the go field name.

liggitt replied:
> Thus I'm thinking we could also fix the go field name EndpointsName to be Endpoints as part of the breaking change. 

leaving the go field names the same means a client that obtains a PV object from client.Core().PersistentVolumes().Get("...") would not need to make changes

humblec commented:
Eventhough the verify works [1] at my local branch, `verify` test fails above with error `I0723 06:33:46.163] hack/make-rules/../../hack/verify-openapi-spec.sh
I0723 06:33:46.163] Makefile:128: recipe for target 'verify' failed
W0723 06:33:46.264] make: *** [verify] Error 1```

[1]
````+++ [0723 15:15:59] Clean up complete
diffing /home/hchiramm/gopath/src/k8s.io/kubernetes/api/openapi-spec against freshly generated openapi spec
/home/hchiramm/gopath/src/k8s.io/kubernetes/api/openapi-spec up to date.
```

humblec commented:
/retest

humblec commented:
/retest

humblec commented:
@liggitt the tests are passed now,  ptal. Thanks!

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -902,6 +902,21 @@ func validateGlusterfsVolumeSource(glusterfs *core.GlusterfsVolumeSource, fldPat
+	}
+	if glusterfs.EndpointsNamespace != nil {
+		endpointNs := glusterfs.EndpointsNamespace
+		allErrs = append(allErrs, field.Required(fldPath.Child("endpointsNamespace"), ""))
```
this should only be an error `if glusterfs.EndpointsNamespace == ""`... as is, this makes providing a namespace an unconditional error

liggitt replied:
that also means we need a test for this :)

humblec replied:
Yeah :),  changed as suggested and added test.

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -160,8 +158,33 @@ func (plugin *glusterfsPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volu
+func (plugin *glusterfsPlugin) getEndpointNameAndNamespace(spec *volume.Spec, defaultNamespace string) (string, string, error) {
+	if spec.Volume != nil && spec.Volume.Glusterfs != nil {
+		endpoints := spec.Volume.Glusterfs.EndpointsName
+		if endpoints != "" {
```
either return `"", defaultNamespace, nil` here or a "no glusterFS endpoint specified" error, falling through to the "Spec does not reference a GlusterFS volume type" error is not correct

humblec replied:
Done

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -160,8 +158,33 @@ func (plugin *glusterfsPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volu
+		endpoints := spec.PersistentVolume.Spec.Glusterfs.EndpointsName
+		endpointsNs := defaultNamespace
+
+		if overriddenNs := spec.PersistentVolume.Spec.Glusterfs.EndpointsNamespace; overriddenNs != nil && len(*overriddenNs) > 0 {
```
validation ensures this is non-empty if non-nil... if set to "", return an error here.

humblec replied:
Done.

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -811,10 +834,11 @@ func (p *glusterfsVolumeProvisioner) CreateVolume(gid int) (r *v1.GlusterfsVolum
-		ReadOnly:      false,
+	return &v1.GlusterfsPersistentVolumeSource{
+		EndpointsName:      endpoint.Name,
+		EndpointsNamespace: &epNamespace,
```
do you plan to adjust the provisioner to allow specifying the endpoint namespace as an option, so it doesn't have to be exposed to the PVC user?

humblec replied:
As of now, getting the option from storage class is not planned, I can avoid exposing it to pvc user while bringing the change. Is it fine?

liggitt replied:
follow up is fine, though it would be ideal to give the option in the same release the PV capability is added to avoid confusion about availability

rootfs replied:
@humblec external provisioner too?

liggitt commented on pkg/volume/glusterfs/glusterfs_test.go:
```diff
@@ -160,7 +160,7 @@ func TestPluginPersistentVolume(t *testing.T) {
 		Spec: v1.PersistentVolumeSpec{
 			PersistentVolumeSource: v1.PersistentVolumeSource{
-				Glusterfs: &v1.GlusterfsVolumeSource{EndpointsName: "ep", Path: "vol", ReadOnly: false},
+				Glusterfs: &v1.GlusterfsPersistentVolumeSource{EndpointsName: "ep", Path: "vol", ReadOnly: true},
```
why did readonly flip here?

liggitt replied:
move the readonly change into its own PR. need to think through the implications more carefully, and want to keep this change purely mechanical

liggitt replied:
ReadOnly should remain true since the readonly behavior change was moved out of this PR

humblec replied:
Done.

liggitt commented on pkg/volume/glusterfs/glusterfs_test.go:
```diff
@@ -181,7 +181,7 @@ func TestPersistentClaimReadOnlyFlag(t *testing.T) {
 		Spec: v1.PersistentVolumeSpec{
 			PersistentVolumeSource: v1.PersistentVolumeSource{
-				Glusterfs: &v1.GlusterfsVolumeSource{EndpointsName: "ep", Path: "vol", ReadOnly: false},
+				Glusterfs: &v1.GlusterfsPersistentVolumeSource{EndpointsName: "ep", Path: "vol", ReadOnly: true},
```
same question about readonly changing

humblec replied:
same as above.

humblec replied:
Done.

liggitt commented on pkg/apis/core/types.go:
```diff
@@ -921,6 +921,29 @@ type GlusterfsVolumeSource struct {
+
+	// EndpointsNamespace is the namespace that contains Glusterfs endpoint.
+	// If this field is empty, the EndpointNamespace defaults to the same namespace as the bound PVC.
+	// More info: https://releases.k8s.io/HEAD/examples/volumes/glusterfs/README.md#create-a-pod
```
add `+optional`

humblec replied:
Done.

ramyala commented:
@humblec can this be slated in for a release milestone?

humblec commented:
@ramyala Targetting this PR for `1.13`.  I will rebase and address final set of comments soon.

humblec commented:
/retest

humblec commented:
@liggitt Review comments are addressed. Please take a look at in your next free slot :), meanwhile I will check why tests are failing. Thanks !!!

humblec commented:
/retest

humblec commented:
All tests are passed. Cheers :+1:

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -903,6 +903,24 @@ func validateGlusterfsVolumeSource(glusterfs *core.GlusterfsVolumeSource, fldPat
+		if *endpointNs == "" {
+			allErrs = append(allErrs, field.Required(fldPath.Child("endpointsNamespace"), ""))
+		} else {
+			allErrs = append(allErrs, ValidateDNS1123Label(*endpointNs, fldPath.Child("endpointsNamespace"))...)
```
prefer `apimachineryvalidation.ValidateNamespaceName`

liggitt commented on pkg/apis/core/validation/validation.go:
```diff
@@ -903,6 +903,24 @@ func validateGlusterfsVolumeSource(glusterfs *core.GlusterfsVolumeSource, fldPat
+	if glusterfs.EndpointsNamespace != nil {
+		endpointNs := glusterfs.EndpointsNamespace
+		if *endpointNs == "" {
+			allErrs = append(allErrs, field.Required(fldPath.Child("endpointsNamespace"), ""))
```
it's not required, an empty string is invalid, so use field.Invalid() with a message indicating if the field is set, it must be a valid namespace name

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -160,8 +158,40 @@ func (plugin *glusterfsPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volu
+		if endpoints != "" {
+			return endpoints, defaultNamespace, nil
+		} else {
+			return "", defaultNamespace, fmt.Errorf("no glusterFS endpoint specified")
```
return `"", "", fmt.Errorf(...)`

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -403,15 +433,15 @@ func (b *glusterfsMounter) setUpAtInternal(dir string) error {
 	} else if spec.PersistentVolume != nil &&
 		spec.PersistentVolume.Spec.Glusterfs != nil {
-		return spec.PersistentVolume.Spec.Glusterfs, spec.ReadOnly, nil
+		return spec.PersistentVolume.Spec.Glusterfs.EndpointsName, spec.PersistentVolume.Spec.Glusterfs.Path, spec.PersistentVolume.Spec.Glusterfs.ReadOnly, nil
```
don't change the readonly behavior as part of this PR. please open that separate as a bugfix and get close eyes on the change from @kubernetes/sig-storage-pr-reviews ... need to think through whether that is compatible or not

liggitt commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -636,6 +636,29 @@ type GlusterfsVolumeSource struct {
+
+	// EndpointsNamespace is the namespace that contains Glusterfs endpoint.
+	// More info: https://releases.k8s.io/HEAD/examples/volumes/glusterfs/README.md#create-a-pod
+	//+optional
```
`// +optional`

liggitt commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -636,6 +636,29 @@ type GlusterfsVolumeSource struct {
+	// +optional
+	ReadOnly bool `json:"readOnly,omitempty" protobuf:"varint,3,opt,name=readOnly"`
+
+	// EndpointsNamespace is the namespace that contains Glusterfs endpoint.
```
match the documentation on the internal type (this one is what actually gets published as API doc)

humblec commented:
/test pull-kubernetes-verify

humblec commented:
/retest

humblec commented:
/retest

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -98,15 +98,17 @@ func (plugin *glusterfsPlugin) GetPluginName() string {
-	volumeSource, _, err := getVolumeSource(spec)
+	var endpointName string
+
+	if spec.Volume != nil && spec.Volume.Glusterfs != nil {
```
this doesn't appear to handle PV volumes correctly

humblec replied:
Added pv volumes to the func.

liggitt commented on pkg/volume/glusterfs/glusterfs.go:
```diff
@@ -403,15 +433,15 @@ func (b *glusterfsMounter) setUpAtInternal(dir string) error {
 }
 
-func getVolumeSource(spec *volume.Spec) (*v1.GlusterfsVolumeSource, bool, error) {
+func getVolumeInfo(spec *volume.Spec) (string, bool, error) {
```
doc what the return values are as well

humblec replied:
sure, done.

humblec commented:
@liggit Thanks!, addressed last set of comments.

humblec commented:
/retest

humblec commented:
@liggit hopefully last commit has addressed return from `GetVolumeName()` as suggested. PTAL .. Thanks !!

humblec commented:
/retest

liggitt commented on vendordiff.patch:
```diff
@@ -1,25 +0,0 @@
-diff -Naupr -x BUILD -x 'AUTHORS*' -x 'CONTRIBUTORS*' vendor/k8s.io/apiserver/pkg/storage/etcd/testing/utils.go /tmp/gopath.qtqLRB/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/storage/etcd/testing/utils.go
```
what is this file?

humblec replied:
Thats a leftover from https://github.com/kubernetes/kubernetes/blob/7f23a743e8c23ac6489340bbb34fa6f1d392db9d/hack/verify-godeps.sh#L109.  The latest `gitignore` has it. But, was not present in my tree, so got added in between. Trying to remove it from index :)

humblec replied:
Done.

liggitt commented:
API lgtm, go ahead and squash in the review commits

needs ack from @kubernetes/sig-storage-pr-reviews on the GetVolumeName() change

liggitt commented:
linting error as well:

```
pkg/volume/glusterfs/glusterfs.go:125:10: if block ends with a return statement, so drop this else and outdent its block
```

msau42 commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -191,7 +191,7 @@ type PersistentVolumeSource struct {
 	// More info: https://releases.k8s.io/HEAD/examples/volumes/glusterfs/README.md
 	// +optional
-	Glusterfs *GlusterfsVolumeSource `json:"glusterfs,omitempty" protobuf:"bytes,4,opt,name=glusterfs"`
+	Glusterfs *GlusterfsPersistentVolumeSource `json:"glusterfs,omitempty" protobuf:"bytes,4,opt,name=glusterfs"`
```
Sorry I'm late to the party. Is this a backwards compatible change?  Will clients that programatically interact with the API start failing because the type has changed?

liggitt replied:
The type of the go field does not impact what gets serialized to the wire, so the type change does not affect wire compatibility (against the API or against etcd).

The type of the go struct field is changing. Depending on how a client made use of this object, it might have to adjust how it constructed a struct of this type. If all it did was retrieve objects from the API and access fields (e.g. `pv.Spec.Glusterfs.EndpointsName`) no change would be required.

msau42 replied:
Ah ok, so if someone synced their go-client to the latest, then it's possible they will have to recompile, but older applications using the old clients will continue to work.  Is this something we need to call out in any release notes?

liggitt replied:
yes, worth a release note, both for the new capability, and for the client-go impact

humblec commented:
/retest

humblec committed:
Refactor GlusterFS PV spec.

This patch introduces glusterfsPersistentVolumeSource addition
to glusterfsVolumeSource. All fields remains same as glusterfsVolumeSource
with an addition of a new field
called `EndpointsNamespace` to define namespace of endpoint in the
spec.

Signed-off-by: Humble Chirammal <hchiramm@redhat.com>

humblec commented:
>API lgtm, go ahead and squash in the review commits>

Thanks a lot @liggitt ! The commits are squashed and test are green :)..

humblec commented:
/retest

humblec commented:
/retest

liggitt commented:
/lgtm
/approve

/hold
for sig-storage approval of GetVolumeName() change. unhold at will once that is acked.

msau42 commented:
/approve

liggitt commented:
/hold cancel
